### ISSUE TITLE
Reinitializable iterator doesn't use cached dataset upon reinitializing

### ISSUE THREAD

**Issue by saxenarohan97** [tensorflow/tensorflow#20930]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes, written custom code
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 8.1
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**: 3.6.5
- **Bazel version**: N/A
- **CUDA/cuDNN version**: CUDA 9.0, cuDNN v7.0
- **GPU model and memory**: NVIDIA GeForce GTX 960 (faced the same problem with 1080 Ti)
- **Exact command to reproduce**: see below

### Describe the problem
Found a likely bug when using a reinitialiable iterator with a cached dataset. Upon reinitializing the iterator, the iterator does not use the cached dataset. This can be verified using the sample code below.

### Source code
```
import tensorflow as tf
import numpy as np
import time

LENGTH = 3
CACHE = True
VERBOSE = True
SHAPE = (1,)

def data_generator():    
    for n in range(LENGTH):
        data = np.random.rand(*SHAPE)
        yield data

def get_dataset():
    train_data = tf.data.Dataset.from_generator(data_generator, output_types=tf.float32)
    if CACHE:
        train_data = train_data.cache()
    return train_data

def get_dataset_repeat():
    train_data = tf.data.Dataset.from_generator(data_generator, output_types=tf.float32)
    if CACHE:
        train_data = train_data.cache()
    train_data = train_data.repeat()
    return train_data

data = get_dataset()
data_repeat = get_dataset_repeat()

iterator = tf.data.Iterator.from_structure(data.output_types, data.output_shapes)
next_element = iterator.get_next()

data_init = iterator.make_initializer(data)
data_init_repeat = iterator.make_initializer(data_repeat)

sess = tf.Session()

print('Reading values from iterator by reinitializing')
epoch = 1

for _ in range(5):
    start = time.time()
    sess.run(data_init)
    print('Epoch %d:' % epoch, end=' ')
    for _ in range(LENGTH):
        output = sess.run(next_element)
        
        if VERBOSE:
            print(output, end=' ')
        
    end = time.time()
    print("Time taken: %f" % (end - start))

    epoch += 1

print('Reading values from iterator using repeat')
epoch = 1
sess.run(data_init_repeat)

for _ in range(5):
    start = time.time()
    print('Epoch %d:' % epoch, end=' ')
    for _ in range(LENGTH):
        output = sess.run(next_element)
        
        if VERBOSE:
            print(output, end=' ')

    end = time.time()
    print("Time taken: %f" % (end - start))
    
    epoch += 1
```

This code gives the following output:

> Reading values from iterator by reinitializing
> Epoch 1: [0.00544547] [0.19976228] [0.5371114] Time taken: 0.017011
> Epoch 2: [0.97181207] [0.35276905] [0.69020385] Time taken: 0.003002
> Epoch 3: [0.19892913] [0.2760849] [0.8980513] Time taken: 0.003003
> Epoch 4: [0.92679894] [0.5854017] [0.6552748] Time taken: 0.003002
> Epoch 5: [0.38050508] [0.7676437] [0.41214108] Time taken: 0.003002
> Reading values from iterator using repeat
> Epoch 1: [0.28107443] [0.8882484] [0.59744525] Time taken: 0.003003
> Epoch 2: [0.28107443] [0.8882484] [0.59744525] Time taken: 0.003002
> Epoch 3: [0.28107443] [0.8882484] [0.59744525] Time taken: 0.000000
> Epoch 4: [0.28107443] [0.8882484] [0.59744525] Time taken: 0.001001
> Epoch 5: [0.28107443] [0.8882484] [0.59744525] Time taken: 0.001000

In the second case, the same values are printed at every epoch because the iterator uses the cached data. However, in the first case, different values are printed at every epoch - indicating that the iterator does not use the cached values. This can be further verified by looking at the time taken (which will be more prominent with larger data). So in the above code, if I set `LENGTH = 100`, `VERBOSE = False` and `SHAPE = (500, 1000, 3)`, I get the following output:

> Reading values from iterator by reinitializing
> Epoch 1: Time taken: 2.228584
> Epoch 2: Time taken: 2.182549
> Epoch 3: Time taken: 2.175545
> Epoch 4: Time taken: 2.185551
> Epoch 5: Time taken: 2.161536
> Reading values from iterator using repeat
> Epoch 1: Time taken: 2.147525
> Epoch 2: Time taken: 0.263195
> Epoch 3: Time taken: 0.259186
> Epoch 4: Time taken: 0.266189
> Epoch 5: Time taken: 0.268181

In the second case, the first epoch takes considerably longer than the subsequent epochs - understandably, since the further epochs just reuse the cached data. However, in the first case, all the epochs take a long time.

This seems like a bug to me because the reinitializable iterator has [been advocated](https://www.tensorflow.org/guide/datasets#creating_an_iterator) as being useful in precisely such kind of situations - when you have 2 different datasets (say training and validation) and you simply reinitialize the iterator to use any one of the two at a time. However, it seems that if I reinitialize the iterator with a cached dataset, I lose the entire advantage of caching. If I set `CACHE = False` with the large shape mentioned above, I get the following output as expected:

> Reading values from iterator by reinitializing
> Epoch 1: Time taken: 2.317660
> Epoch 2: Time taken: 2.251585
> Epoch 3: Time taken: 2.209581
> Epoch 4: Time taken: 2.151515
> Epoch 5: Time taken: 2.156546
> Reading values from iterator using repeat
> Epoch 1: Time taken: 2.182563
> Epoch 2: Time taken: 2.207575
> Epoch 3: Time taken: 2.205560
> Epoch 4: Time taken: 2.139524
> Epoch 5: Time taken: 2.198561

That is, each epoch takes the same time with either method since we don't cache the dataset.

**Comment by mrry** [comment]
Assigning this one to @saeta, since he is working on a replacement for `Dataset.cache()` that probably ought to support this (especially in eager mode!).

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version

**Comment by saxenarohan97** [comment]
I have updated the description, I'm not sure how to trigger the bot. @shivaniag can you assign the relevant person again - I think it is @saeta as @mrry mentioned. TIA.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shivaniag** [comment]
@saeta do you have an update on this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @saeta: It has been 106 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by arthur-gal** [comment]
@saeta do we have an update on this ?

**Comment by Kwander** [comment]
I am facing the same issue... I want to cache a dataset X which is generates a training set Y with a reinitializable iterator going through X. Every time I reinitialize X iterator, even though I cache X, it iterates X as if there was no caching.

**Comment by rkdx007** [comment]
@saeta @mrry Any solution to this? I am also facing a similar issue.

**Comment by jsimsa** [comment]
In TF 1.X, the state of the cache is tied to the state of the iterator and re-initializing the iterator will recreate the state of the iterator (including the cache state). We do not plan to change the semantics of "reinitialize" w.r.t. to the state of the iterator.

Having said that, I am planning to shortly submit a PR that, in TF 2.0, will decouple the cache state from the iterator state. In TF 2.0, instead of re-initializing an iterator you simply create multiple Python iterators (either implicitly via `for elem in dataset:` or explicitly via `iter(dataset)`. Once the aforementioned PR is submitted, the different iterators will share the same cache.
